solana/programs/token-bridge-relayer/src/processor/outbound.rs

```diff
@@ -0,0 +1,505 @@
+use crate::{
+    constants::{SEED_PREFIX_BRIDGED, SEED_PREFIX_CUSTODY},
+    error::TokenBridgeRelayerError,
+    message::RelayerMessage,
+    state::{
+        calculate_total_fee, ChainConfigAccount, ForeignContractAccount, SignerSequenceAccount,
+        TbrConfigAccount,
+    },
+    KiloLamports, TargetChainGas,
+};
+use anchor_lang::prelude::*;
+use anchor_spl::token::{Mint, Token, TokenAccount};
+use solana_price_oracle::state::{EvmPricesAccount, PriceOracleConfigAccount};
+use wormhole_anchor_sdk::{
+    token_bridge::{self, program::TokenBridge},
+    wormhole::program::Wormhole,
+};
+
+/// Transfer a native SPL token.
+///
+/// The other side will mint a wrapped token issued by Wormhole.
+/// The other side will unbound a native ERC-20 token.
+#[derive(Accounts)]
+#[instruction(
+    recipient_chain: u16,
+    recipient_address: [u8; 32],
+)]
+pub struct OutboundTransfer<'info> {
+    /// Payer will pay Wormhole fee to transfer tokens and create temporary
+    /// token account.
+    #[account(mut)]
+    pub payer: Signer<'info>,
+
+    /// This program's config.
+    #[account(
+        has_one = fee_recipient @ TokenBridgeRelayerError::WrongFeeRecipient,
+        seeds = [TbrConfigAccount::SEED_PREFIX],
+        bump
+    )]
+    pub tbr_config: Account<'info, TbrConfigAccount>,
+
+    /// The peer config. We need to verify that the transfer is sent to the
+    /// canonical peer.
+    #[account(
+        constraint = chain_config.canonical_peer_address == recipient_address
+            @ TokenBridgeRelayerError::MustSendToCanonicalPeer,
+        seeds = [
+            ChainConfigAccount::SEED_PREFIX,
+            recipient_chain.to_be_bytes().as_ref(),
+        ],
+        bump
+    )]
+    pub chain_config: Account<'info, ChainConfigAccount>,
+
+    /// Foreign Contract account. Send tokens to the contract specified in this
+    /// account. Funnily enough, the Token Bridge program does not have any
+    /// requirements for outbound transfers for the recipient chain to be
+    /// registered. This account provides extra protection against sending
+    /// tokens to an unregistered Wormhole chain ID. Read-only.
+    #[account(
+        seeds = [
+            ForeignContractAccount::SEED_PREFIX,
+            &recipient_chain.to_be_bytes()
+        ],
+        bump,
+    )]
+    pub foreign_contract: Account<'info, ForeignContractAccount>,
```

Comment:
```suggestion
```
If we want to preserve the check, then the token bridge address field should be moved to the chain config.